### PR TITLE
ICA certificates in Credential Issuer

### LINKED ISSUE #7724: Implement generating ICA certificates in Credential Issuer
#### Problem
The credential issuers (`ExampleOperationalCredentialsIssuer`, and `CHIPOperationalCredentialsDelegate`) do not generate ICA certificates. Due to this, `CHIPDeviceController` is currently commissioning devices with NOC signed by Root CAs.

We need to add the support for ICA so that the complete flow can be exercised.

 #### Proposed Solution
1. Implement `GetIntermediateCACertificate` in `ExampleOperationalCredentialsIssuer`, and `CHIPOperationalCredentialsDelegate` classes.
2. Sign the NOC with ICA keypair (instead of Root's keypair)
3. Test that commissioner is able to successfully pair with the device.

### PR DESCRIPTION
**Pull request by jpk233**
#### Problem
The credential issuers (ExampleOperationalCredentialsIssuer, and CHIPOperationalCredentialsDelegate) do not generate ICA certificates. Due to this, CHIPDeviceController is currently commissioning devices with NOC signed by Root CAs.

Fixes #7724

#### Change overview
* Implemented GetIntermediateCACertificate in ExampleOperationalCredentialsIssuer.

* Added new method to decode CHIP Certificate Arrays and convert them into separate CHIP certificates (NOC and ICA).

* Added ICA Certificate Support to AdminPairingTable.

* Renamed OperationalCert argument to NOCArray

#### Testing
* Added two tests to TestChipCert.cpp for the certificate array decode method described above.
* Tested commissioning flow using chip-tool and lighting-app apps.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -261,6 +261,14 @@ CHIP_ERROR DeviceController::LoadLocalCredentials(Transport::AdminPairingInfo *
 
         ReturnErrorOnFailure(admin->SetOperationalCert(ByteSpan(chipCert, chipCertLen)));
 
+        ChipLogProgress(Controller, "Getting ICA certificate for the controller from the issuer");
+        ReturnErrorOnFailure(mOperationalCredentialsDelegate->GetIntermediateCACertificate(0, cert, kMaxCHIPOpCertLength, certLen));
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:265]: It is not mandatory to have an ICAC in the chain. The path needs to account for optionality of the ICAC here. That is, there should be a way to determine absence from `GetIntermediateCACertificate` and skip the conversion and set the ICA to empty with `SetICACert`

File: src/credentials/CHIPCert.h [hunk 2]
```diff
@@ -653,6 +653,24 @@ CHIP_ERROR ConvertX509CertToChipCert(const uint8_t * x509Cert, uint32_t x509Cert
...
+ * @brief Convert CHIP certificate array to Separate NOC and ICA Certificates.
+ *        This function takes a CHIP certificate array and splits it into single
+ *        NOC and ICA CHIP certificates.
+ *
+ *        NOC certificate must be provided.
+ *        ICA certificate is optional. It will be omitted (nullptr, 0) if the CHIP certificate array contains only one entry.
+ *
+ *        The API enforces that the NOC is issued by ICA (if ICA is provided).
+ *
+ * @param operationalCert Chip certificate array.
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert.h:666]: Suggest renaming to opCertArray, since it's not just a single certificate.

File: src/credentials/CHIPCertFromX509.cpp [hunk 3]
```diff
@@ -765,5 +765,67 @@ CHIP_ERROR ConvertX509CertsToChipCertArray(const ByteSpan & x509NOC, const ByteS
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR ConvertChipCertArrayToChipCerts(const ByteSpan & operationalCert, ByteSpan & noc, ByteSpan & icac)
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCertFromX509.cpp:768]: This does not convert, it extracts. Suggest renaming `ExtractCertsFromCertArray`

File: src/transport/AdminPairingTable.cpp [hunk 4]
```diff
@@ -232,6 +243,44 @@ CHIP_ERROR AdminPairingInfo::SetRootCert(const ByteSpan & cert)
...
+    }
+
+    if (mICACert == nullptr)
+    {
+        mICACert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(cert.size()));
+    }
+    VerifyOrReturnError(mICACert != nullptr, CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(CanCastTo<uint16_t>(cert.size()), CHIP_ERROR_INVALID_ARGUMENT);
+    mICACertLen          = static_cast<uint16_t>(cert.size());
+    mICACertAllocatedLen = (mICACertLen > mICACertAllocatedLen) ? mICACertLen : mICACertAllocatedLen;
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.cpp:278]: Is this intended? This sounds like an error if `(mICACertLen > mICACertAllocatedLen)`. How can you have a cert larger than the allocated space. What is the intent of this path? Shouldn't `ReleaseICACert` always set `mICACertAllocatedLen` to 0?

File: src/transport/AdminPairingTable.cpp [hunk 5]
```diff
@@ -232,6 +243,44 @@ CHIP_ERROR AdminPairingInfo::SetRootCert(const ByteSpan & cert)
...
+    {
+        ReleaseICACert();
+    }
+
+    if (mICACert == nullptr)
+    {
+        mICACert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(cert.size()));
+    }
+    VerifyOrReturnError(mICACert != nullptr, CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(CanCastTo<uint16_t>(cert.size()), CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.cpp:276]: This `VerifyOrReturnError(CanCastTo<uint16_t>(cert.size()), CHIP_ERROR_INVALID_ARGUMENT);` needs to be moved *before* the alloc.

File: src/transport/AdminPairingTable.cpp [hunk 6]
```diff
@@ -258,15 +307,24 @@ CHIP_ERROR AdminPairingInfo::SetOperationalCert(const ByteSpan & cert)
         ReleaseOperationalCert();
     }
 
+    ByteSpan noc;
+    ByteSpan icac;
+    ConvertChipCertArrayToChipCerts(cert, noc, icac);
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.cpp:312]: The overall method `SetOperationalCert` should be renamed `SetOperationalCertArray`.

File: src/transport/AdminPairingTable.cpp [hunk 7]
```diff
@@ -258,15 +307,24 @@ CHIP_ERROR AdminPairingInfo::SetOperationalCert(const ByteSpan & cert)
...
+
     if (mOperationalCert == nullptr)
     {
-        mOperationalCert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(cert.size()));
+        mOperationalCert = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(noc.size()));
     }
     VerifyOrReturnError(mOperationalCert != nullptr, CHIP_ERROR_NO_MEMORY);
-    VerifyOrReturnError(CanCastTo<uint16_t>(cert.size()), CHIP_ERROR_INVALID_ARGUMENT);
-    mOpCertLen          = static_cast<uint16_t>(cert.size());
+    VerifyOrReturnError(CanCastTo<uint16_t>(noc.size()), CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.cpp:324]: This check must be move BEFORE the alloc.

Also, on error of NOC alloc, we need to release the ICA cert.

Strongly suggest the following:

1. Refactor `SetOperationalCert` to `SetOperationalCertsFromCertArray`
2. Make a `SetNOCCert` method that is similar to `SetICACert` in construction
3. Use `SetICACert`/`SetNOCCert` in here rather than having the inlined logic of `SetNOCCert`

File: src/transport/AdminPairingTable.cpp [hunk 8]
```diff
@@ -281,6 +339,12 @@ CHIP_ERROR AdminPairingInfo::GetCredentials(OperationalCredentialSet & credentia
         certificates.LoadCert(mRootCert, mRootCertLen,
                               BitFlags<CertDecodeFlags>(CertDecodeFlags::kIsTrustAnchor).Set(CertDecodeFlags::kGenerateTBSHash)));
 
+    if (mICACert != nullptr && mICACertLen > 0)
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.cpp:342]: Comment on the purpose of this block.

File: src/transport/AdminPairingTable.h [hunk 9]
```diff
@@ -167,6 +169,9 @@ class DLL_EXPORT AdminPairingInfo
     uint8_t * mRootCert            = nullptr;
     uint16_t mRootCertLen          = 0;
     uint16_t mRootCertAllocatedLen = 0;
+    uint8_t * mICACert             = nullptr;
+    uint16_t mICACertLen           = 0;
+    uint16_t mICACertAllocatedLen  = 0;
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.h:174]: Why do we care about the allocated lengths? Could we reduce potential state mismatch hazards with just `mICACertLen` and `mOpCertLen` ?

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
Need to fix several error paths related to issues during alloc

File: src/transport/AdminPairingTable.cpp (hunk 4 above)
Comment (jpk233) [reply src/transport/AdminPairingTable.cpp:278]: It's mimicing the SetRootCert method. However I agree that the mICACertAllocatedLen should be removed because of this odd logic.

File: src/transport/AdminPairingTable.cpp (hunk 8 above)
Comment (jpk233) [reply src/transport/AdminPairingTable.cpp:342]: It optionally adds the whole certificate chain (including ICA) if ICA exists.

File: src/transport/AdminPairingTable.cpp (hunk 7 above)
Comment (jpk233) [reply src/transport/AdminPairingTable.cpp:324]: Thanks, will update.

File: src/transport/AdminPairingTable.cpp (hunk 6 above)
Comment (jpk233) [reply src/transport/AdminPairingTable.cpp:312]: Will update.

File: src/transport/AdminPairingTable.cpp (hunk 5 above)
Comment (jpk233) [reply src/transport/AdminPairingTable.cpp:276]: Will update.

File: src/credentials/CHIPCertFromX509.cpp (hunk 3 above)
Comment (jpk233) [reply src/credentials/CHIPCertFromX509.cpp:768]: Will update.

File: src/credentials/CHIPCert.h (hunk 2 above)
Comment (jpk233) [reply src/credentials/CHIPCert.h:666]: Will update.

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (jpk233) [reply src/controller/CHIPDeviceController.cpp:265]: Will update.

File: src/transport/AdminPairingTable.h [hunk 10]
```diff
@@ -190,12 +196,13 @@ class DLL_EXPORT AdminPairingInfo
...
         uint16_t mRootCertLen; /* This field is serialized in LittleEndian byte order */
+        uint16_t mICACertLen;  /* This field is serialized in LittleEndian byte order */
         uint16_t mOpCertLen;   /* This field is serialized in LittleEndian byte order */
 
         Crypto::P256SerializedKeypair mOperationalKey;
         uint8_t mRootCert[kMaxChipCertSize];
-        // The operationa credentials set can have up to two certs -> ICAC and NOC
-        uint8_t mOperationalCert[kMaxChipCertSize * 2];
+        uint8_t mICACert[kMaxChipCertSize];
+        uint8_t mOperationalCert[kMaxChipCertSize];
```
[the remark this answers is not in the document]
Comment (jpk233) [reply src/transport/AdminPairingTable.h:205]: Will update.

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 11]
```diff
@@ -348,7 +348,7 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
     AdminPairingInfo * admin = retrieveCurrentAdmin();
     VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
 
-    VerifyOrExit(admin->SetOperationalCert(OperationalCert) == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(admin->SetOperationalCertsFromCertArray(OperationalCert) == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
```
Comment (tcarmelveilleux) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:351]: Suggest renaming OperationalCert argument to NOCArray. I will open a spec PR to renormalize this, now that OperationalCert argument was renamed NOC.

Comment (tcarmelveilleux) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:351]: See https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3437 for some background

File: src/credentials/CHIPCert.h [hunk 12]
```diff
@@ -653,6 +653,24 @@ CHIP_ERROR ConvertX509CertToChipCert(const uint8_t * x509Cert, uint32_t x509Cert
...
 
+/**
+ * @brief Extract NOC and ICA Certificates from a CHIP certificate array.
+ *        This function takes a CHIP certificate array and splits it into single
+ *        NOC and ICA CHIP certificates.
+ *
+ *        NOC certificate must be provided.
+ *        ICA certificate is optional. It will be omitted (nullptr, 0) if the CHIP certificate array contains only one entry.
+ *
+ *        The API enforces that the NOC is issued by ICA (if ICA is provided).
```
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert.h:664]: How is that enforced? I don't see code to check chaining. That chaining should be validated at time of use, perhaps, and this comment line removed if the check is not added. I think that the check is probably a good check, but not easy to do right now.

File: src/transport/AdminPairingTable.cpp (hunk 8 above)
Comment (tcarmelveilleux) [reply src/transport/AdminPairingTable.cpp:356]: I meant to say "Add a comment on the purpose of this block" :) We reuse the `mICACert != nullptr && mICACertLen > 0` condition a lot. Perhaps it would be good to have `IsICACAvailable` and `IsNOCAvailable` trivial bool const predicates to reduce the mental burden of these pointer/size checks in the code.

File: src/transport/AdminPairingTable.h [hunk 13]
```diff
@@ -190,12 +196,13 @@ class DLL_EXPORT AdminPairingInfo
...
-        uint16_t mOpCertLen;   /* This field is serialized in LittleEndian byte order */
+        uint16_t mICACertLen;  /* This field is serialized in LittleEndian byte order */
+        uint16_t mNOCCertLen;  /* This field is serialized in LittleEndian byte order */
 
         Crypto::P256SerializedKeypair mOperationalKey;
         uint8_t mRootCert[kMaxChipCertSize];
-        // The operationa credentials set can have up to two certs -> ICAC and NOC
-        uint8_t mOperationalCert[kMaxChipCertSize * 2];
+        uint8_t mICACert[kMaxChipCertSize];
+        uint8_t mNOCCert[kMaxChipCertSize];
```
Comment (tcarmelveilleux) [inline src/transport/AdminPairingTable.h:205]: These should all be `kMaxChipCertSize` rather than `kMaxChipCertSize`. Please audit the other places that `kMaxChipCertSize` is used to make sure no call site expect `kMaxChipCertSize` as a constant value, when the intent was sizeof(mNOCCert/mICACert), etc.

File: src/credentials/CHIPCert.h (hunk 12 above)
Comment (jpk233) [reply src/credentials/CHIPCert.h:664]: We do check if the TLV reached the end after parsing the NOC - Check lines 806-812 from CHIPCertFromX509.cpp. If it reaches the end, we set the ICAC as (nullptr, 0)

Comment (tcarmelveilleux) [reply src/credentials/CHIPCert.h:664]: Verifying presence/absence of NOC is not the same of `enforcing that the NOC is issued by ICA`. I see no parsing to validate that the AKID of the NOC matches the SKID of the ICA if present, nor issuer subject check, nor any signature verification of NOC from ICA. I strongly recommand removing this one comment about "enforces that the NOC is issued by ICA", since that check should be done after extracting both certs, during AddNOC/UpdateNOC processing, and not as part of a basic data access/getter operation.

Comment (jpk233) [reply src/credentials/CHIPCert.h]: I removed the comment as per your suggestion.

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp [hunk 14]
```diff
@@ -5318,7 +5318,7 @@ void DispatchServerCommand(app::Command * apCommandObj, CommandId aCommandId, En
         {
         case ZCL_ADD_OP_CERT_COMMAND_ID: {
             expectArgumentCount = 4;
-            chip::ByteSpan OperationalCert;
+            chip::ByteSpan NOCArray;
```
Comment (pan-apple) [inline examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:5321]: This change will get overwritten when someone regenerates the code from the ZAP files.
Also, suggest to keep renaming as a separate PR. It'll help review the core changes in the current PR.

If we want to keep the rename in this PR, we should update the XML and ZAP files, and regenerate the code.

File: src/transport/AdminPairingTable.cpp [hunk 15]
```diff
@@ -232,45 +244,105 @@ CHIP_ERROR AdminPairingInfo::SetRootCert(const ByteSpan & cert)
...
+    if (certArray.size() == 0)
+    {
+        ReleaseNOCCert();
+        ReleaseICACert();
+        return CHIP_NO_ERROR;
+    }
+
+    ByteSpan noc;
+    ByteSpan icac;
+    ExtractCertsFromCertArray(certArray, noc, icac);
```
Comment (pan-apple) [inline src/transport/AdminPairingTable.cpp:330]: We don't need to split the cert array into NOC and ICAC. The CASESession is cable of loading the cert array itself using `LoadCerts` method. Besides that, is there any other reason for splitting it?

**Comment by pan-apple** [comment]
> #### Testing
> * Added two tests to TestChipCert.cpp for the certificate array decode method described above.

Please also test E2E provisioning flow using controller apps. This change will have impact on CASE session establishment.

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp (hunk 14 above)
Comment (tcarmelveilleux) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:5321]: Agree about renaming this one later

File: src/transport/AdminPairingTable.cpp (hunk 15 above)
Comment (tcarmelveilleux) [reply src/transport/AdminPairingTable.cpp:330]: The reason is that this avoids needing to store/allocate the contiguous block. No certificate chain should be stored as a linear chain. They appear on the wire as an array, but they don't need to follow through everywhere as an array. Keeping them separate allows eventually simplifying the storage arrangement to truly not require linear storage of ICAC + NOC.

**Review by tcarmelveilleux — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by jpk233** [comment]
> > #### Testing
> > 
> > * Added two tests to TestChipCert.cpp for the certificate array decode method described above.
> 
> Please also test E2E provisioning flow using controller apps. This change will have impact on CASE session establishment.

Sorry, failed to include this in the testing section before. This has also been E2E tested similar to our upcoming CASE refresh. This has been added now.

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp (hunk 14 above)
Comment (jpk233) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:5321]: We're updating the operational-credentials-cluster.xml and ZAP files in this PR already to address the name change.

File: src/transport/AdminPairingTable.h (hunk 13 above)
Comment (jpk233) [reply src/transport/AdminPairingTable.h]: Updated.

File: src/transport/AdminPairingTable.cpp (hunk 15 above)
Comment (bzbarsky-apple) [reply src/transport/AdminPairingTable.cpp:330]: But if we always want them on the wire as a single blob, why would we not store them as that blob so we can blit directly onto the wire?

Otherwise we have to either transcode or do something where we store raw TLV bytes and insert them into a TLV we are constructing or something like that, right?

**Review by hawk248 — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

File: src/controller/ExampleOperationalCredentialsIssuer.cpp [hunk 16]
```diff
@@ -88,14 +107,28 @@ ExampleOperationalCredentialsIssuer::GenerateNodeOperationalCertificate(const Op
     ReturnErrorCodeIf(!noc.Alloc(kMaxCHIPDERCertLength), CHIP_ERROR_NO_MEMORY);
     uint32_t nocLen = 0;
 
-    ReturnErrorOnFailure(NewNodeOperationalX509Cert(request, CertificateIssuerLevel::kIssuerIsRootCA, pubkey, mIssuer, noc.Get(),
-                                                    kMaxCHIPDERCertLength, nocLen));
+    ReturnErrorOnFailure(NewNodeOperationalX509Cert(request, CertificateIssuerLevel::kIssuerIsIntermediateCA, pubkey,
```
Comment (pan-apple) [inline src/controller/ExampleOperationalCredentialsIssuer.cpp:110]: The `mIssuerId` in the `request` should be changed to `mIntermediateIssuer`. Otherwise, the NOC issuer will be set to the root.

Comment (jpk233) [reply src/controller/ExampleOperationalCredentialsIssuer.cpp:110]: Thanks, updated.
